hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java

```diff
@@ -381,13 +381,16 @@ private XceiverClientReply sendCommandWithRetry(
       }
     }
     if (datanodeList == null) {
-      if (topologyAwareRead) {
-        datanodeList = pipeline.getNodesInOrder();
-      } else {
+      List<DatanodeDetails> orderedNodes =
+          topologyAwareRead ? pipeline.getNodesInOrder() : null;
+      // if node order is not defined - use random shuffled order
+      if (orderedNodes == null || orderedNodes.isEmpty()) {
         datanodeList = pipeline.getNodes();
         // Shuffle datanode list so that clients do not read in the same order
         // every time.
         Collections.shuffle(datanodeList);
+      } else {
+        datanodeList = orderedNodes;
```

Comment:
I think this should be cleaner.
```suggestion
      datanodeList =
          topologyAwareRead ? pipeline.getNodesInOrder() : null;
      // if node order is not defined - use random shuffled order
      if (datanodeList == null || datanodeList.isEmpty()) {
        datanodeList = pipeline.getNodes();
        // Shuffle datanode list so that clients do not read in the same order
        // every time.
        Collections.shuffle(datanodeList);
      } 
```

Reply:
Right, thanks!